### PR TITLE
server: limit concurrency of table stats requests

### PR DESCRIPTION
**Pull request by ajwerner**
These requests can be very expensive, especially if there are lots of ranges. Furthermore, the number of requests to be sent tends to scale with the number of tables, just like the range count. This unbounded concurrency was a recipe for trouble.

Informs #90196.

Release note (ui change): Requests to fetch table and database statistics now have limited concurrency. This may make loading the database page slower, but in return should result in making those pages less disruptive.

### CODE REVIEW

**Comment by ajwerner** [comment]
@knz can I get your take on this?

**Review by knz — COMMENTED** [review]
This is not horrible, but the primordial sin here is that the api endpoint is badly designed.
It should handle pagination (as we've specced out in the /api/v2 tree). If not using pagination, it should also use a memory budget.

Your PR is not making the problem worse but it's papering over the misdesign.

I'd like a TL from SQL Obs to talk with Obs Infra on how to improve this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-:-NElSnLlDG0ZSAX67zCi:b-jcos8f)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by ajwerner** [comment]
I agree with the need for pagination. I also think we need something better than what we have in terms of the efficiency of the underlying requests (https://github.com/cockroachdb/cockroach/issues/84105), but, for now, I think we need to do something. As it stands, for a cluster with 10,000 tables in 10 databases, we'll launch 40011 goroutines and send 10000 requests concurrently to each node. It can totally DoS a cluster. It happens to be the case that also each request is O(N) with regards to replicas, and was ridiculously inefficient.  

Even if we set the limit to something more conservative like 128, I think this is definitely worth doing.

**Comment by knz** [comment]
Agreed, I have no objection.

**Comment by knz** [comment]
should still get a review from obs inf though.

**Review by matthewtodd — COMMENTED** [review]
Wow, thanks, @ajwerner. Left a couple of questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-:-NElfxOYDjEhSiF6Z48l:bh0c7qz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[`pkg/server/admin.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NElbdhwA4OGaUNadcr5:-NElbdhwA4OGaUNadcr6:b-s23gb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L138)):*
> ```Go
> 		}
> 		return nil
> 	},
> ```

FWIW, there is [`settings.PositiveInt`](https://github.com/cockroachdb/cockroach/blob/b39f02aaf2a0ea4fe93a26a22d5c4d82943a8340/pkg/settings/int.go#L152-L158) to do this same validation.
___
*[`pkg/server/admin.go` line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEld6-lBE7kEXRFUnxy:-NEld6-lBE7kEXRFUnxz:bvdlazy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L617)):*
> ```Go
> 			ctx, stop.TaskOpts{
> 				TaskName: "server.adminServer: requesting table stats",
> 				Sem:      sem,
> ```

IIUC, with this change, `getDatabaseStats` will [return an error](https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/util/stop/stopper.go#L426-L454) rather than running more than 32 concurrent table stats requests.

Did you want to set `WaitForSem` in the `TaskOpts`?
___
*[`pkg/server/admin.go` line 1306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NElfNrk4NSwaEEuZ2iv:-NEleN07BmfCYbboG2uY:b-ru41u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L1306)):*
> ```Go
> 				TaskName: "server.adminServer: requesting remote stats",
> 				Sem:      s.statsLimiter,
> 			},
> ```

Ditto on the `WaitForSem` question here.

**Comment by ajwerner** [comment]
Good catch on `WaitForSem`!

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @matthewtodd)

___
*[`pkg/server/admin.go` line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleMMEDgHHe7_ppc4c:-NEleMMEDgHHe7_ppc4d:bbcr5jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L133)):*
> ```Go
> 	"maximum number of concurrent table stats fetches to run",
> 	32, // arbitrary
> 	func(i int64) error {
> ```

`settings.PositiveInt,`
___
*[`pkg/server/admin.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleuP5-_8FmZkewO0t:-NEleuP5-_8FmZkewO0u:b-e8mt1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L151)):*
> ```Go
> 		server:                s,
> 	}
> 	if cs := s.ClusterSettings(); cs != nil { // can be nil in testing
> ```

can really `s != nil` and `s.ClusterSettings() == nil`? I've looked superficially at which test this is talking about and couldn't figure it out.
___
*[`pkg/server/admin.go` line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NElfLPs9NeHVXc0IjwD:-NElfLPs9NeHVXc0IjwE:b-fiuu4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab412341ca1ae0c2d5c2a76a35ccf1c967a2ebb5/pkg/server/admin.go#L583)):*
> ```Go
> 			func(ctx context.Context) {
> 				statsResponse, err := s.statsForSpan(ctx, tableSpan)
> ```

consider not deleting this line if you go back to the file
___
*[`pkg/server/admin.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleYf1EA4DBXNvS1_i:-NEleYf1EA4DBXNvS1_j:b-93g3bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L600)):*
> ```Go
> 
> 	// There's no reason to launch more goroutines than we will ultimately
> 	// be able to launch
> ```

The repetition in this comment is no good. Also missing period at the end.
But I had to stare at this for a while until I figured out what this is trying to tell me. Please put more words in pointing to the limit inside `statsForSpan`, and say that the two levels can't share a semaphore in order to avoid deadlock.

--- 

I think the nice way to build this would be for the lower-level to use a semaphore with a dynamic capacity. The capacity would be 1 + whatever it can pull from a higher-level parent semaphore. We can approximate that with a "semi-sync" scheme - make the lower level not use `WaitForSem`, and run each task synchronously on error, and then try `RunAsyncTaskEx()` again on the next loop iteration.
___
*[`pkg/server/admin.go` line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEld6-lBE7kEXRFUnxy:-NElgTK8EBtF3Aj1f2Fu:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L617)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

IIUC, with this change, `getDatabaseStats` will [return an error](https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/util/stop/stopper.go#L426-L454) rather than running more than 32 concurrent table stats requests.

Did you want to set `WaitForSem` in the `TaskOpts`?
</blockquote></details>

+1

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @matthewtodd)

___
*[`pkg/server/admin.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleYf1EA4DBXNvS1_i:-NEllEBq5yyTFqXu_4mQ:b-y5u6hi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L600)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The repetition in this comment is no good. Also missing period at the end.
But I had to stare at this for a while until I figured out what this is trying to tell me. Please put more words in pointing to the limit inside `statsForSpan`, and say that the two levels can't share a semaphore in order to avoid deadlock.

--- 

I think the nice way to build this would be for the lower-level to use a semaphore with a dynamic capacity. The capacity would be 1 + whatever it can pull from a higher-level parent semaphore. We can approximate that with a "semi-sync" scheme - make the lower level not use `WaitForSem`, and run each task synchronously on error, and then try `RunAsyncTaskEx()` again on the next loop iteration.

</blockquote></details>

 >  We can approximate that with a "semi-sync" scheme - make the lower level not use WaitForSem, and run each task synchronously on error, and then try RunAsyncTaskEx() again on the next loop iteration.

And then the kicker is that the two levels would share the one and only semaphore - `s.statsLimiter`.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @matthewtodd)

___
*[`pkg/server/admin.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleuP5-_8FmZkewO0t:-NElmW229cv2QQ1s5tit:bkqsw7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L151)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can really `s != nil` and `s.ClusterSettings() == nil`? I've looked superficially at which test this is talking about and couldn't figure it out.
</blockquote></details>

I promise there were panics, I didn't look where. 
___
*[`pkg/server/admin.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleYf1EA4DBXNvS1_i:-NEllzdn-axUgx7XCPDi:brzlpyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L600)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

 >  We can approximate that with a "semi-sync" scheme - make the lower level not use WaitForSem, and run each task synchronously on error, and then try RunAsyncTaskEx() again on the next loop iteration.

And then the kicker is that the two levels would share the one and only semaphore - `s.statsLimiter`.
</blockquote></details>

To what extent do you think this complexity is needed now? I'd rather avoid it for backport and let somebody else do something more involved.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @matthewtodd)

___
*[`pkg/server/admin.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleuP5-_8FmZkewO0t:-NElomLz2oFa1_eq9FIS:b3j9vmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L151)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I promise there were panics, I didn't look where. 
</blockquote></details>

well, let's look a little bit :P
Cause it'd be nice to not have these nil guards.
And I also have to confess some curiosity, since I don't see this `newAdminServer` called directly, and the callers that exist I think give it a proper `Server`.
___
*[`pkg/server/admin.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleYf1EA4DBXNvS1_i:-NElodRZBlUQid-yG9yj:brxe89) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L600)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

To what extent do you think this complexity is needed now? I'd rather avoid it for backport and let somebody else do something more involved. 
</blockquote></details>

Up to you. All I insist on is a better comment.

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-:-NEr9BkzDIl-IsROU9Bm:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @matthewtodd)

___
*[`pkg/server/admin.go` line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleMMEDgHHe7_ppc4c:-NEr423x4AT3dMiPCDl6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L133)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`settings.PositiveInt,`
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleuP5-_8FmZkewO0t:-NEr7Sd25GMsgitao_xe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L151)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well, let's look a little bit :P
Cause it'd be nice to not have these nil guards.
And I also have to confess some curiosity, since I don't see this `newAdminServer` called directly, and the callers that exist I think give it a proper `Server`.
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NElfLPs9NeHVXc0IjwD:-NEr7_941fskwy-KpWGX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab412341ca1ae0c2d5c2a76a35ccf1c967a2ebb5/pkg/server/admin.go#L583)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider not deleting this line if you go back to the file
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEleYf1EA4DBXNvS1_i:-NEr7TNMDIl1svP2xxHS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fc9fbf11a3d82cb9db9fd77745ed245ce58df61/pkg/server/admin.go#L600)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Up to you. All I insist on is a better comment.
</blockquote></details>

Done.

**Review by andreimatei — APPROVED** [review]
:lgtm:  but please check if there's another bug in that ctor

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-:-NEviEbW1lHjpUFfBDgw:b-u2e7ow)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @matthewtodd)

___
*[`pkg/server/admin.go` line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvhplxAJpE-waAABZo:-NEvhplxAJpE-waAABZp:bacitts) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L137)):*
> ```Go
> 
> // newAdminServer allocates and returns a new REST server for administrative
> // APIs. Note that the server here is not yet initialized and cannot be used
> ```

It took me a second to understand what's going on there. Consider saying `s` instead of "server" - otherwise it's ambiguous with the "admin server" being returned. And consider saying that `s` will be initialized by the time the RPCs served by the admin server are called.
___
*[`pkg/server/admin.go` line 147 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvgbGx8l3YW87SOxIa:-NEvgbGx8l3YW87SOxIb:b-exv8ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L147)):*
> ```Go
> 		server:                s,
> 	}
> 	server.statsLimiter = quotapool.NewIntPool(
> ```

why isn't this field assignment inside the struct literal above?
___
*[`pkg/server/admin.go` line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvgCq3DacUC2fSNuCQ:-NEvgCq3DacUC2fSNuCR:bta85z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L156)):*
> ```Go
> 		)
> 	})
> 	//}
> ```

some garbage here
___
*[`pkg/server/admin.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvh3JuE7djetmNIv44:-NEvh3JuE7djetmNIv45:biorra2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L167)):*
> ```Go
> 		nil,
> 		noteworthyAdminMemoryUsageBytes,
> 		s.ClusterSettings(),
> ```

so is this sucker busted too?
___
*[`pkg/server/server.go` line 748 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEviB9jAxHzT09mE32J:-NEviB9jAxHzT09mE32K:bkc0wx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/server.go#L748)):*
> ```Go
> 	adminAuthzCheck := &adminPrivilegeChecker{ie: internalExecutor, st: st, makePlanner: nil}
> 	sAdmin := newAdminServer(
> 		lateBoundServer, cfg.Settings, adminAuthzCheck, internalExecutor,
> ```

thanks for doing something

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-:-NEvujio-onJ5kCOHNJz:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei and @matthewtodd)

___
*[`pkg/server/admin.go` line 137 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvhplxAJpE-waAABZo:-NEvmlNi2Fn9IfzSdNIx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L137)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It took me a second to understand what's going on there. Consider saying `s` instead of "server" - otherwise it's ambiguous with the "admin server" being returned. And consider saying that `s` will be initialized by the time the RPCs served by the admin server are called.
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 147 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvgbGx8l3YW87SOxIa:-NEvmyVM533at_6zZQMg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L147)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why isn't this field assignment inside the struct literal above?
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvgCq3DacUC2fSNuCQ:-NEvmz1h6kcqPEvcoFyi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L156)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

some garbage here
</blockquote></details>

Done.
___
*[`pkg/server/admin.go` line 167 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90210#-NEvh3JuE7djetmNIv44:-NEvmzqG1bR7aJJJMehs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8b7139779548f82880801cc1d111d757fe83886/pkg/server/admin.go#L167)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

so is this sucker busted too?
</blockquote></details>

Done.

**Comment by ajwerner** [comment]
blathers backport 22.2.0

**Comment by ajwerner** [comment]
blathers backport 22.2

**Comment by ajwerner** [comment]
blathers backport 22.1

**Pull request merged**
